pkg/registry/client.go

```diff
@@ -27,99 +29,100 @@ type tokenResponse struct {
 }
 
 type Client interface {
-	Manifest(repository, reference string) (distribution.Manifest, string, error)
+	GetBearerToken() (string, error)
+	GetManifest(bearerToken string) (distribution.Manifest, error)
 }
 
 type client struct {
-	registryURL   string
-	client        *http.Client
-	authorization string
+	scanRequest harbor.ScanRequest
+	client      *http.Client
 }
 
 type clientFactory struct {
+	tlsConfig etc.TLSConfig
 }
 
-func (cf *clientFactory) Get(registryURL, authorization string) (Client, error) {
+func NewClientFactory(TLSConfig etc.TLSConfig) ClientFactory {
+	return &clientFactory{
+		tlsConfig: TLSConfig,
+	}
+}
+
+func (cf *clientFactory) Get(scanRequest harbor.ScanRequest) (Client, error) {
 	return &client{
-		registryURL: registryURL,
+		scanRequest: scanRequest,
 		client: &http.Client{Transport: &http.Transport{
 			TLSClientConfig: &tls.Config{
-				// FIXME Allow configuring custom or self-signed certs rather than skipping verification.
-				InsecureSkipVerify: true,
+				RootCAs:            cf.tlsConfig.RootCAs,
+				InsecureSkipVerify: cf.tlsConfig.InsecureSkipVerify,
 			},
 		}},
-		authorization: authorization,
 	}, nil
 }
 
-func NewClientFactory() ClientFactory {
-	return &clientFactory{}
-}
-
-func (c *client) Manifest(repository, reference string) (distribution.Manifest, string, error) {
-	tokenRequest, err := c.getTokenRequest(repository, reference)
-	if err != nil {
-		return nil, "", err
-	}
-	log.Debugf("Token request: %v", tokenRequest)
-	tokenResponse, err := c.getAccessToken(tokenRequest)
-	if err != nil {
-		return nil, "", err
-	}
-
-	requestURL := fmt.Sprintf("%s/v2/%s/manifests/%s", c.registryURL, repository, reference)
-	log.Debugf("Fetch manifest URL: %s", requestURL)
-	req, err := http.NewRequest("GET", requestURL, nil)
+func (c *client) GetManifest(bearerToken string) (distribution.Manifest, error) {
+	requestURL := c.manifestURL()
+	log.WithField("url", requestURL).Debug("Fetching manifest started")
+	req, err := http.NewRequest(http.MethodGet, requestURL, nil)
 	if err != nil {
-		return nil, "", err
+		return nil, err
 	}
 
-	req.Header.Add(http.CanonicalHeaderKey("Accept"), schema2.MediaTypeManifest)
-	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", tokenResponse.Token))
+	req.Header.Add("Accept", schema2.MediaTypeManifest)
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", bearerToken))
 
 	resp, err := c.client.Do(req)
 	if err != nil {
-		return nil, "", err
+		return nil, err
 	}
-	log.Debugf("Response status: %s", resp.Status)
-	log.Debugf("Response headers: %v", resp.Header)
+	log.WithField("status", resp.Status).Debug("Fetching manifest finished")
 
 	defer resp.Body.Close()
 	b, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return nil, "", err
+		return nil, err
 	}
 
 	if resp.StatusCode != http.StatusOK {
-		return nil, "", fmt.Errorf("HTTP not ok: %v", resp.Status)
+		return nil, fmt.Errorf("HTTP not ok: %v", resp.Status)
```

Comment:
Is there any error message in the response body `b`, which can be appended to the error obj?

Reply:
If there sth in the response body, I'm appending it to the error message.